crates/nu-command/src/commands/generators/hash_/generic_digest.rs

```diff
@@ -0,0 +1,58 @@
+use crate::prelude::*;
+use nu_errors::ShellError;
+use nu_protocol::ShellTypeName;
+use nu_protocol::{ColumnPath, Primitive, UntaggedValue, Value};
+use nu_source::Tag;
+
+pub fn run<D>(args: CommandArgs) -> Result<OutputStream, ShellError>
+where
+    D: digest::Digest,
+    digest::Output<D>: core::fmt::LowerHex,
+{
+    let column_paths: Vec<ColumnPath> = args.rest(0)?;
+
+    Ok(args
+        .input
+        .map(move |v| {
+            if column_paths.is_empty() {
+                action::<D>(&v, v.tag())
+            } else {
+                let mut ret = v;
+
+                for path in &column_paths {
+                    ret = ret.swap_data_by_column_path(
+                        path,
+                        Box::new(move |old| action::<D>(old, old.tag())),
+                    )?;
+                }
+
+                Ok(ret)
+            }
+        })
+        .into_input_stream())
+}
+
+pub fn action<D>(input: &Value, tag: Tag) -> Result<Value, ShellError>
+where
+    D: digest::Digest,
+    digest::Output<D>: core::fmt::LowerHex,
+{
+    match &input.value {
+        UntaggedValue::Primitive(Primitive::String(s)) => {
+            let digest_result = D::digest(s.as_bytes());
+            Ok(UntaggedValue::string(&format!("{:x}", digest_result)).into_value(tag))
+        }
+        UntaggedValue::Primitive(Primitive::Binary(bytes)) => {
+            let digest_result = D::digest(bytes);
+            Ok(UntaggedValue::string(&format!("{:x}", digest_result)).into_value(tag))
+        }
+        other => {
+            let got = format!("got {}", other.type_name());
+            Err(ShellError::labeled_error(
+                "value is not supported for hashing",
```

Comment:
This error message used to be "value is not supported for hashing as md5", now it's more generic since it's used by all hashers. Not sure if this is a problem or not.

Reply:
it would be nice if it identified the hashing algorithm that isn't supported. i wonder if there's an easy way to determine that here.

Reply:
The trait itself doesn't provide any `name()` or similar but it should be easy to add one. My thinking was that the behaviour regarding input values is identical for all algorithms so it wouldn't add much information for the user.

That said, I was looking at implementing common `usage()`, `signature()` and perhaps `run()` all reusing the `name()` but that'd result in additional changes to `WholeStreamCommand::usage()` since it returns `&str`. If it's fine I can adapt it to return `Cow`. I mean I can make the change and if it's too invasive we don't have to merge it.

Reply:
let's land this one and if you find a good way to add the name then that's fine too. thanks for your contribution!